[SPIR-V]: Memory leak fix in SPIRVEmitIntrinsics

bwlodarcz opened the pull request:
The architecture of SPIRVEmitIntrinsics is build in such way that every private method is called by one main function runOnFunction which then calls private methods. Private member IRB is allocated in runOnFunction method but it's not freed. Due to that every time when IR function contains intrinsics to emit, runOnFunction is entered and memory is leaked on exit. It's especially true when there are two or more IR functions to emit. IRB is set to nullptr during construction of object and it's left without pointing resource until runOnFunction is entered. This also create possibility of simple mistake when private method is called but there is no resource pointed. Change requires passing IRBuilder by reference to private methods. The visit* functions create it's own IRBuilder thus IRB is eliminated from class scope. In addition there is a small performance improvement because IRBuilder is not allocated by heap.

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:
```diff
@@ -660,33 +669,34 @@ bool SPIRVEmitIntrinsics::runOnFunction(Function &Func) {
+    insertPtrCastInstr(I, B);
   }
-
+  this->IRB = &B;
```
Thank you for the patch and catching the issue! Indeed, looks like we were not freeing IRB. However, I am wondering, would it be possible to either use only the B as the reference by passing it to pass' other methods or assign this->IRB early on in the runOnFunction and only use this approach. I think only the latter approach will be possible. 

Right now the methods we override rely on the IRB pointer, while the not overriding methods have B passed as a reference. Do you think it would be better to stick to just one approach?

bwlodarcz replied:
Passing B everywhere would be the best option but unfortunately we can't use `B` as the reference in whole function because visit* methods have strict interface and changing them fails to compile definition files.
The problem with `this->IRB` approach is that lifecycle of the resource is not visible from the interface of the functions.
The class logic flow looks like that:
1. enter `runOnFunction`
2. assign pointer to IRB
3. call private helper methods
4. run visit
5. exit `runOnFunction`

The flow isn't visible in class and raises questions instantaneously when approached:  
1. When IRB is created?
2. Can I use private methods before I run `runOnFunction`? - no
3. When IRB is again a `nullptr`?
4. How the class is used? Isn't IRB pointer passed somewhere to deeper frame or returned to higher?

The uncertain lifecycle of IRB was IMO the actual cause of the error. 
By changing interfaces of the private methods to require the reference sends visible signal to everybody reading the code that IRBuilder needs to exist to call private method.
I think that clearer fix would be to add support class which will take visit* interfaces on itself. That PR would be likely bigger than this one.

michalpaszkowski replied:
Thank you for elaborating and looking into this carefully! I agree, this might be a better approach from the readability standpoint.

michalpaszkowski reviewed (approved):
LGTM! I will wait for other comments by the end of this week. Otherwise, assuming LITs and CTS pass, I will merge this change.

bwlodarcz committed:
[SPIR-V]: Memory leak fix in SPIRVEmitIntrinsics

The architecture of SPIRVEmitIntrinsics is build in
such way that every private method is called by one
main function runOnFunction which then calls private methods.
Private member IRB is allocated in runOnFunction method but it's not freed.
Due to that every time when IR function contains intrinsics to emit,
runOnFunction is entered and memory is leaked on exit.
It's especially true when there are two or more IR functions to emit.
IRB is set to nullptr during construction of object
and it's left without pointing resource until runOnFunction is entered.
This also create possibility of simple mistake when
private method is called but there is no resource pointed.
Change requires passing IRBuilder by reference to private methods.
The visit* functions create it's own IRBuilder instance.
In addition there is a small performance improvement
because IRBuilder is not allocated by heap.

bwlodarcz committed:
Fix format
